src/models/TeamGenerator.ts

```diff
@@ -75,47 +75,102 @@ export default class TeamGenerator {
     public static generate(players: Array<Player>, teamSize: number ,game: Game): Array<Team> | GeneratorErrorCode{
 
         //Step 1: do proper param check
-            
-        //--> at least 2 full teams need to be createable: players.length >= 2* teamSize
-        if(players.length < 2 * teamSize){
-            return GeneratorErrorCode.TeamSizeAndPlayerLengthMismatch;
+        let paramCheckResult: GeneratorErrorCode | undefined = this.validateGeneratorInput(players, teamSize, game);
+        if (paramCheckResult !== undefined){
+            return paramCheckResult;
         }
-        //--> each player must have a valid assessment for the given game
-        for (const player of players){
-            if(!player.isSkillAssessedForGame(game)){
-                return GeneratorErrorCode.PlayerSkillsIncomplete;
+
+        //Step 2: order players by their skill descending and shuffle players with same skill in array
+        let orderedPlayerArray: Array<Player> = this.orderPlayersDescendingByGameSkill(players, game);
+
+        
+        //Step 3: create X full teams and one additional team with remaining players if any
+        let [fullTeams, additionalTeam, avgPlayerSkill] = this.assignPlayersToTeams(orderedPlayerArray, teamSize, game);
+        
+
+        //Step 4: Refine team balance (swap players between best and worst team if possible)
+        // fill up additional team with fake substitution player of avg skill for refinement step, remove afterwards
+        if (additionalTeam.fixedPlayers.length > 0){
+            for(let i = additionalTeam.currentSize; i < additionalTeam.targetSize; i++){
+                let player: Player = new Player("FakeSubPlayer" + i);
+                player.addGameSkill(new GameSkill(game, avgPlayerSkill));
             }
+            fullTeams.push(additionalTeam); // add team with sub players to collection for balance optimizations
         }
 
-        //Step 2: order players by their skill ascending and shuffle players with same skill in array
-        // group players with same skill
+        // optimize balance between teams
+        this.optimizeTeamSkillBalance(fullTeams, game);
+
+        // remove fake sub players again from additional team:
+        additionalTeam.clearSubstitutionPlayers();
+
+        return fullTeams;
+    }
+
+    /**
+     * 
+     * @param players 
+     * @param teamSize 
+     * @param game 
+     * @returns 
```

Comment:
Do you want to add the return type? e.g. @returns {(GeneratorErrorCode|undefined)}

Reply:
Hm... at least in VSCode I do see the return type only in the signature when the function description is displayed. @return does not show the type. I copy pasted your suggestion as it is.